proposals/0327-actor-initializers.md

```diff
@@ -42,126 +49,129 @@ The protection is achieved by _isolating_ the mutable state of each actor instan
-  3. *Minor:* Initializer delegation requires the use of the `convenience` keyword like classes, even though actors do not support inheritance. Is this attribute still needed? If so, what is the isolation of an actor's convenience initializer?
+```swift
+actor Clicker {
+  var count: Int
+  func click() { self.count += 1 }
+  nonisolated func announce() { print("performing a click!") }
+
+  init(ok: Void) {
+    self.count = 0
+    Task { await self.click() }
+    self.announce() // rejected in Swift 5.5, but is race-free.
+  }
+}
+```
 
-The following subsections will discuss these three high-level problems in more detail.
+Leveraging the actor isolation model, we know `announce` cannot touch the stored property `count` to observe that `click` happened concurrently with the initializer. That's because `announce` is not isolated to the actor instance. In fact, a race can only happen in the initializer if an access to `count` appears after the creation of the task. This proposal aims to generalize that idea into something we refer to _flow-sensitive isolation_, which uses [data-flow analysis](https://en.wikipedia.org/wiki/Data-flow_analysis) to prove statically that non-async initializers of an actor are race-free.
 
-### Initializer Races
+### Data-races in deinitializers
 
-Unlike other synchronous methods of an actor, a synchronous (or "ordinary") `init` is special in that it is treated as being `nonisolated` from the outside, meaning that there is no `await` (or actor hop) required to call the `init`. This is because an `init`'s purpose is to bootstrap a fresh actor-instance, called `self`. Thus, at various points within the `init`'s body, `self` is considered a fully-fledged actor instance whose members must be protected by isolation. The existing implementation of actor initializers does not perform this enforcement, leading to data races with the code appearing in the `init`:
+While non-async initializers gained restrictions to prevent data races in Swift 5.5, deinitializers did not. Yet, `deinit`s can still exhibit data races and illegal lifetime extensions of `self` that live longer than the `deinit`'s invocation. One kind of data race in a `deinit` is conceptually the same as the one described earlier for non-async initializers:
 
 ```swift
-actor StatsTracker {
-  var counter: Int
+actor Clicker {
+  var count: Int = 0
 
-  init(_ start: Int) {
-    self.counter = start
-    // -- self fully-initialized here --
-    Task.detached { await self.tick() }
-    
-    // ... do some other work ...
-    
-    if self.counter != start { // 💥 race
-      fatalError("state changed by another thread!")
+  func click(_ times: n) {
+    for _ in 0<..n {
+      self.count += 1 
     }
   }
 
-  func tick() {
-    self.counter = self.counter + 1
+  deinit {
+    let current = count
+    
+    Task { await self.click(1000000) } // This might keep `self` alive after the `deinit` too!
```

Comment:
"Might" is true I guess but can we word this as undefined behavior instead (we do so below in text)?